File: 0946-validate-stack-sequences/0946-validate-stack-sequences.py
class Solution:
    def validateStackSequences(self, pushed: List[int], popped: List[int]) -> bool:
        stack = []
        i = 0
        j = 0
        n = len(pushed)
        while j < n:
            if stack and (stack[-1] == popped[j]):
                stack.pop()
                j += 1
            elif i < n:
                stack.append(pushed[i])
                i += 1
            else:
                return False
        return True

File: 0946-validate-stack-sequences/0946-validate-stack-sequences.rs
impl Solution {
    pub fn validate_stack_sequences(pushed: Vec<i32>, popped: Vec<i32>) -> bool {
        let mut stack: Vec<i32> = Vec::new();
        let mut i: usize = 0;
        let mut j: usize = 0;
        let n: usize = pushed.len() as usize;
        let mut stack_len: usize;
        while (j < n) {
            if (!stack.is_empty()) {
                stack_len = stack.len() as usize;
                if (stack[stack_len-1] == popped[j]) {
                    stack.pop();
                    j += 1;
                }
                else if (i < n) {
                    stack.push(pushed[i]);
                    i += 1;
                }
                else {
                    return false;
                }
            }
            else if (i < n) {
                stack.push(pushed[i]);
                i += 1;
            }
            else {
                return false;
            }
        }
        return true;
    }
}
